text/0821-public-types.md

```diff
@@ -0,0 +1,358 @@
+---
+Stage: Accepted
+Start Date: 2022-05-23
+Release Date: Unreleased
+Release Versions:
+  ember-source: vX.Y.Z
+  ember-data: vX.Y.Z
+Relevant Team(s):
+  - Ember.js
+  - TypeScript
+  - Learning
+RFC PR: https://github.com/emberjs/rfcs/pull/821
+---
+
+# Public API for Type-Only Imports
+
+
+## Summary
+
+Introduce public import locations for type-only imports which have previously had no imports, and fully specify their public APIs for end users:
+
+- `Owner`, with `TypeOptions` and `FactoryManager`
+- `Transition`
+- `RouteInfo` and `RouteInfoWithAttributes`
+
+
+## Motivation
+
+Prior to supporting TypeScript, Ember has defined certain types as part of its API, but *without* public imports, since the types were not designed to be imported, subclassed, etc. by users. For the community-maintained type definitions, the Typed Ember team chose to match that policy so as to avoid committing the main project to public API. With the introduction of TypeScript as a first-class language (see esp. RFCs [0724: Official TypeScript Support][0724] and [0800: TypeScript Adoption Plan][0800]) those types now need public imports so that users can reference them; per the [Semantic Versioning for TypeScript Types spec][spec], they also need to define *how* they are public: can they be sub-classed, re-implemented, etc.?
+
+[0274]: https://rfcs.emberjs.com/id/0724-road-to-typescript
+[0800]: https://rfcs.emberjs.com/id/0800-ts-adoption-plan
+[spec]: https://www.semver-ts.org
+
+Additionally, the lack of a public import or contract for the `Owner` interface has been a long-standing problem for *all* users, but especially TypeScript users, and given the APIs provided for e.g. the Glimmer Component class where `owner` is the first argument, the pervasive use of `getOwner` in low-level library code, etc., it is important for TypeScript users to be able to use an `Owner` safely, and for JavaScript users to be able to get autocomplete etc. from the types we ship.
+
+
+## Detailed design
+
+**Note:** For the terms "user-constructible" and "user-subclassable" see the [Semantic Versioning for TypeScript Types spec][spec].
+
+### `Owner`
+
+`Owner` is a user-constructible interface, with an intentionally-minimal subset of the :
+
+```ts
+export default interface Owner {
+  lookup(name: string): unknown;
+
+  register<T>(
+    fullName: string,
+    factory: FactoryManager<T>,
+    options?: TypeOptions
+  ): void;
+
+  hasRegistration(fullName: string): boolean;
+
+  resolveRegistration(
+    fullName: string
+  ): FactoryManager<unknown> | object | undefined;
+}
+```
+
+It is the default import from a new module, `@ember/owner`. With it come two other new types: `TypeOptions` and `FactoryManager`.
+
+#### `TypeOptions`
```

Comment:
Is this actually a good name?

Reply:
This is just me, but if it comes from the `@ember/owner` module/package, my thought is that we should just call it `RegisterOptions` and let the inherent namespace effect of modules do its job (and if folks want to explicitly name it that way they can). There are (at least per ripgrep!) no *other* `RegisterOptions` anywhere within Ember so we don't even have the possible auto-import overlap issue there.

Reply:
Sounds good (and still better than `TypeOptions`, which sounds way too generic)